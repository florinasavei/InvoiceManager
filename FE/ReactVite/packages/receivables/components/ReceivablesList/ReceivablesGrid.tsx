import { useEffect, useState } from "react"
import { IReceivableDTO } from "@invoice-manager/models"
import {
    DataGridRowParams,
    DataGridValueFormatterParams,
    DataGridVisibleColumns,
    MaterialDataGrid,
    MaterialDataGridCol
} from "@invoice-manager/data-grid"
import {
    BooleanIndicator,
    FormattedCurrency,
    WidthToggler,
    calculateRemainingPercentage,
    formatISODate,
    useWindowSize
} from "@invoice-manager/core"
import { ReceivablesGridLegend } from "./ReceivablesGridLegend"

import styles from './style.module.scss'

interface IReceivablesGridProps {
    receivables: IReceivableDTO[] | undefined
}

export const ReceivablesGrid = ({ receivables }: IReceivablesGridProps): React.ReactElement => {

    const [width] = useWindowSize();
    const [isSmallView, setIsSmallView] = useState<boolean>(width < 1200);

    useEffect(() => {
        setIsSmallView(width < 1200)
    }, [width])

    const getRowStyle = (params: DataGridRowParams<IReceivableDTO>) => {
        if (params.row.Cancelled === true) {
            return styles["ReceivablesGrid__Row--canceled"];
        }
        return "";
    };

    const columns: Array<MaterialDataGridCol<IReceivableDTO>> = [
        {
            field: 'UniqueReference',
            headerName: "Unique Reference",
            sortable: true,
            flex: 3
        },
        {
            field: 'OpeningValue',
            headerName: "Opening Value",
            sortable: true,
            flex: 3,
            renderCell: (data: DataGridRowParams<IReceivableDTO>) => {
                const value = data.row.OpeningValue;
                const currencyCode = data.row.CurrencyCode;
                return <div title={`${currencyCode} ${value}`} className={styles['ReceivablesGrid__Cell--ellipsis']}>
                    <FormattedCurrency value={value} currencyCode={currencyCode} />
                </div>
            }
        },
        {
            field: 'RemainingBalance',
            headerName: "Remaining Balance",
            sortable: true,
            flex: 3,
            renderCell: (data: DataGridRowParams<IReceivableDTO>) => {
                const value = data.row.RemainingBalance;
                const currencyCode = data.row.CurrencyCode;
                return <div title={`${currencyCode} ${value}`} className={styles['ReceivablesGrid__Cell--ellipsis']}>
                    <FormattedCurrency value={value} currencyCode={currencyCode} />
                </div>
            }
        },
        {
            field: 'ComputedDeptPercentage',
            headerName: "Dept",
            sortable: true,
            flex: 2,
            valueGetter: (params: DataGridRowParams<IReceivableDTO>) =>
                calculateRemainingPercentage(params.row.RemainingBalance, params.row.OpeningValue),
            valueFormatter: (params: DataGridValueFormatterParams<number>) => `${params.value} %`
        },
        {
            field: 'IssueDate',
            headerName: "Issue Date",
            sortable: true,
            flex: 4,
            renderCell: (data: DataGridRowParams<IReceivableDTO>) => {
                const value = formatISODate(data.row.IssueDate.toString());
                return <div title={value} className={styles['ReceivablesGrid__Cell--ellipsis']}>
                    {value}
                </div>
            }
        },
        {
            field: 'DueDate',
            headerName: "Due Date",
            sortable: true,
            flex: 4,
            renderCell: (data: DataGridRowParams<IReceivableDTO>) => {
                const value = formatISODate(data.row.DueDate.toString());
                return <div title={value} className={styles['ReceivablesGrid__Cell--ellipsis']}>
                    {value}
                </div>
            }
        },
        {
            field: 'ClosedDate',
            headerName: "Closed Date",
            sortable: true,
            flex: 4,
            renderCell: (data: DataGridRowParams<IReceivableDTO>) => {
                if (data.row.ClosedDate === null) {
                    return '-';
                }
                const value = formatISODate(data.row.ClosedDate.toString());
                return <div title={value} className={styles['ReceivablesGrid__Cell--ellipsis']}>
                    {value}
                </div>
            }
        },
        {
            field: 'Cancelled',
            headerName: "Active",
            sortable: true,
            flex: 1,
            valueGetter: (params: DataGridRowParams<IReceivableDTO>) => !params.row.Cancelled,
            renderCell: (data: DataGridRowParams<IReceivableDTO>) =>
                <BooleanIndicator value={!data.row.Cancelled} />
        },
    ]

    let visibleColumns: DataGridVisibleColumns = {
        ComputedDeptPercentage: !isSmallView,
        ClosedDate: !isSmallView,
        IssueDate: !isSmallView,
        Cancelled: !isSmallView,
    }

    return (
        <div test-dataid="ReceivablesGrid">
            {receivables == undefined || !(receivables.length > 0) ?
                <>No Data</>
                :
                <>
                    <WidthToggler isExpanded={!isSmallView} toggleExpand={(expanded) => { setIsSmallView(!expanded) }} />
                    <MaterialDataGrid
                        columns={columns}
                        rows={receivables}
                        rowIdField='Id'
                        getRowStyle={getRowStyle}
                        columnVisibility={visibleColumns}
                        gridId={`Receivables-${isSmallView ? 'narrow' : 'wide'}`}
                    />
                    <ReceivablesGridLegend />
                </>
            }
        </div>
    )
}
