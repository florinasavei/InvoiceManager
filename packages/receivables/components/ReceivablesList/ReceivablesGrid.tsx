import { useEffect, useLayoutEffect, useState } from "react"
import { IReceivableDTO } from "@invoice-manager/models"
import {
    DataGridRowParams,
    DataGridValueFormatterParams,
    DataGridVisibleColumns,
    MaterialDataGrid,
    MaterialDataGridCol
} from "@invoice-manager/data-grid"
import {
    BooleanIndicator,
    FormattedCurrency,
    WidthToggler,
    calculateRemainingPercentage
} from "@invoice-manager/core"

import styles from './style.module.scss'


interface IReceivablesGridProps {
    receivables: IReceivableDTO[] | undefined
}

// TODO: move to core
function useWindowSize() {
    const [size, setSize] = useState([0, 0]);
    useLayoutEffect(() => {
        function updateSize() {
            setSize([window.innerWidth, window.innerHeight]);
        }
        window.addEventListener('resize', updateSize);
        updateSize();
        return () => {
            window.removeEventListener('resize', updateSize)
        };
    }, []);
    return size;
}

export const ReceivablesGrid = ({ receivables }: IReceivablesGridProps): React.ReactElement => {

    const [width] = useWindowSize();

    const [isSmallView, setIsSmallView] = useState<boolean>(width < 1200);


    useEffect(() => {
        setIsSmallView(width < 1200)
    }, [width])

    const getRowStyle = (params: DataGridRowParams<IReceivableDTO>) => {
        if (params.row.Cancelled === true) {
            return styles["ReceivablesGrid__Row--canceled"];
        }
        return "";
    };

    const columns: Array<MaterialDataGridCol<IReceivableDTO>> = [
        {
            field: 'UniqueReference',
            headerName: "Unique Reference",
            sortable: true,
            flex: 1
        },
        {
            field: 'OpeningValue',
            headerName: "Opening Value",
            sortable: true,
            flex: 1,
            renderCell: (data: DataGridRowParams<IReceivableDTO>) =>
                <FormattedCurrency value={data.row.OpeningValue} currencyCode={data.row.CurrencyCode} />
        },
        {
            field: 'RemainingBalance',
            headerName: "Remaining Balance",
            sortable: true,
            flex: 1,
            renderCell: (data: DataGridRowParams<IReceivableDTO>) =>
                <FormattedCurrency value={data.row.RemainingBalance} currencyCode={data.row.CurrencyCode} />
        },
        {
            field: 'ComputedDeptPercentage',
            headerName: "Dept",
            sortable: true,
            flex: 1,
            valueGetter: (params: DataGridRowParams<IReceivableDTO>) =>
                calculateRemainingPercentage(params.row.RemainingBalance, params.row.OpeningValue),
            valueFormatter: (params: DataGridValueFormatterParams<number>) => `${params.value} %`
        },
        {
            field: 'IssueDate',
            headerName: "Issue Date",
            sortable: true,
            flex: 1
        },
        {
            field: 'DueDate',
            headerName: "Due Date",
            sortable: true,
            flex: 1
        },
        {
            field: 'ClosedDate',
            headerName: "Closed Date",
            sortable: true,
            flex: 1,
        },
        {
            field: 'Cancelled',
            headerName: "Active",
            sortable: true,
            flex: 1,
            valueGetter: (params: DataGridRowParams<IReceivableDTO>) => !params.row.Cancelled,
            renderCell: (data: DataGridRowParams<IReceivableDTO>) =>
                <BooleanIndicator value={!data.row.Cancelled} />
        },
    ]

    let visibleColumns: DataGridVisibleColumns = {
        ComputedDeptPercentage: !isSmallView,
        ClosedDate: !isSmallView,
        IssueDate: !isSmallView,
        Cancelled: !isSmallView,
    }

    return (
        <div>
            <WidthToggler isExpanded={!isSmallView} toggleExpand={(expanded) => { setIsSmallView(!expanded) }} />
            {receivables &&
                <MaterialDataGrid
                    columns={columns}
                    rows={receivables}
                    rowIdField='Id'
                    getRowStyle={getRowStyle}
                    columnVisibility={visibleColumns}
                    gridId={`Receivables-${isSmallView ? 'narrow' : 'wide'}`}
                />}
        </div>
    )
}
