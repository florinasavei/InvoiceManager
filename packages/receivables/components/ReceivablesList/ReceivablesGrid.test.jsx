import React from 'react';
import { render } from '@testing-library/react';
import { ReceivablesGrid } from './ReceivablesGrid';

const mockReceivableDTO = {
    Id: 1,
    UniqueReference: 'ABC123',
    OpeningValue: 1000,
    RemainingBalance: 500,
    ComputedDeptPercentage: 50,
    IssueDate: '2023-10-29',
    DueDate: '2023-11-29',
    ClosedDate: '2023-12-29',
    Cancelled: false,
    CurrencyCode: 'USD',
};

// Mock the useWindowSize hook
jest.mock('@invoice-manager/core', () => ({
    useWindowSize: jest.fn(() => [1200]), // Mock the width as 1200 by default
    WidthToggler: ({ isExpanded, toggleExpand }) => <></>
}));

jest.mock('@invoice-manager/data-grid', () => ({
    MaterialDataGrid: ({ columns,
        rows,
        rowIdField,
        getRowStyle,
        columnVisibility,
        gridId
    }) => <></>
}));


jest.mock('@invoice-manager/models', () => ({
}));

describe('ReceivablesGrid', () => {
    it('renders without errors', () => {
        const receivables = [mockReceivableDTO];

        const { container } = render(<ReceivablesGrid receivables={receivables} />);

        expect(container).toBeInTheDocument();
    });
});

test('Checked component matches snapshot', () => {
    const receivables = [mockReceivableDTO];

    const { asFragment } = render(<ReceivablesGrid receivables={receivables} />);
    expect(asFragment()).toMatchSnapshot();
  });
  