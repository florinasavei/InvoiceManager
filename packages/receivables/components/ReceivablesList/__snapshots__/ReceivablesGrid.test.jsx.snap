// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checked component matches snapshot 1`] = `
<DocumentFragment>
  <div />
</DocumentFragment>
`;
